tensorflow/compiler/mlir/tosa/transforms/legalize_tfl_stateful.cc

```diff
@@ -0,0 +1,190 @@
+        // Quantized types we can not use the m_Constant matcher.
+        if (auto constOp = dyn_cast<mlir::TFL::QConstOp>(
+                assign.getValue().getDefiningOp())) {
+          constant = constOp.getValue().cast<DenseElementsAttr>();
+        }
+      }
+      if (!constant) continue;
+
+      // Create TOSA VariableOps
+      auto name = handle.getSharedName();
+      auto global = builder.create<mlir::tosa::VariableOp>(
+          handle.getLoc(), name, constant.getType(), constant);
+      symbolRefMap[name] = global;
+    }
+  }
+  // TF::CallOnceOps are no longer needed as we have already extracted their
+  // state.
+  for (auto op : callOnceOps) op.erase();
+
+  // Replace the assign ops with a tosa store operation.
+  for (auto assign : assignOps) {
+    auto handle = dyn_cast<mlir::TFL::VarHandleOp>(
+        assign.getResourceId().getDefiningOp());
+    if (!handle) continue;
+
+    Value value = assign.getValue();
+    auto globalOpIt = symbolRefMap.find(handle.getSharedName());
+    if (globalOpIt == symbolRefMap.end()) {
+      assign->emitError(
+          "Unable to find corresponding TosaOp for op's VarHandle");
+      continue;
+    }
+    auto globalOp = std::get<1>(*globalOpIt);
+
+    builder.setInsertionPoint(assign);
+    if (globalOp.getType() != value.getType()) {
+      value = builder
+                  .create<UnrealizedConversionCastOp>(assign.getLoc(),
+                                                      globalOp.getType(), value)
+                  .getResult(0);
+    }
+
+    builder.create<mlir::tosa::VariableWriteOp>(
+        assign.getLoc(), llvm::StringRef(globalOp.getName()), value);
+    assign.erase();
+  }
+
+  for (auto read : readOps) {
+    auto handle =
+        dyn_cast<mlir::TFL::VarHandleOp>(read.getResourceId().getDefiningOp());
+    if (!handle) continue;
+
+    auto globalOpIt = symbolRefMap.find(handle.getSharedName());
+    if (globalOpIt == symbolRefMap.end()) continue;
+    auto globalOp = std::get<1>(*globalOpIt);
+
+    builder.setInsertionPoint(read);
+
+    Value load = builder.create<mlir::tosa::VariableReadOp>(
+        read.getLoc(), globalOp.getType(), llvm::StringRef(globalOp.getName()));
+
+    if (read.getType() != load.getType()) {
+      load = builder
+                 .create<UnrealizedConversionCastOp>(read.getLoc(),
+                                                     read.getType(), load)
+                 .getResult(0);
+    }
+    read.getResult().replaceAllUsesWith(load);
+    read.erase();
+  }
+
+  for (auto handle : handleOps) {
+    if (handle.getResult().use_empty()) {
+      handle.erase();
+    }
+  }
+}
+
+}  // namespace
+
+// Creates an instance of the TensorFlow Lite dialect LegalizeTFLStateful pass.
+std::unique_ptr<OperationPass<ModuleOp>> createLegalizeTFLStatefulPass() {
+  return std::make_unique<TosaLegalizeTFLStateful>();
+}
+
+}  // namespace tosa
+}  // namespace mlir
```

Comment:
New line at end

Reply:
Ack. It's strange that I pushed the file with new line at the end. but I can't see it in the Github UI...Not sure what you saw on another side?